nise/yaml_generators/azure/generator.py

```diff
@@ -77,43 +69,62 @@ def generate_instance_id(key, config):
     return f"subscriptions/{config.payer_account}/resourceGroups/{resource_group}/{accts_str[1:-2]}/{resource_name}"
 
 
-def generate_tags(key, config, prefix="", suffix="", dynamic=True, _random=False):
+def generate_tags_and_instance_id(key, config, prefix="", suffix="", dynamic=True):
     """Generate properly formatted Azure tags.
     Returns:
         list
     """
-    keys = TAG_KEYS.get(key)
-    if _random:
-        keys = random.sample(keys, k=random.randint(1, len(keys)))
-    return [dicta(key=key, v=generate_name(config)) for key in keys]
-
-
-def generate_azure_dicta(config, key, _random):
+    if not config.get("id_labels"):
+        keys = TAG_KEYS.get(key)
+        tags = [dicta(key=key, v=generate_name(config)) for key in keys]
+        instance_id = generate_instance_id(key, config)
+    else:
+        id_label_key = random.choice(list(config.id_labels.keys()))
+        tag_key_list = random.choice(config.id_labels.get(id_label_key))
+        SEEN_KEYS = set()
+        tags = []
+        for key, value in tag_key_list:
+            if key not in SEEN_KEYS:
+                tags.append(dicta(key=key, v=value))
+                SEEN_KEYS.update([key])
+        _, node_name = id_label_key
+        instance_id = generate_instance_id(key, config, node_name)
+    return tags, instance_id
```

Comment:
Does the `instance_id` need to be generated here? It seems like it'd be cleaner to have two methods: `generate_instance_id()` and `generate_tags()` rather than putting them together.

What do you think?

Reply:
Initially, I had them separate, but the confusing bit comes from trying to make ocp-on-azure yamls. I need the node-name and corresponding tags from OCP and I settled on a dictionary with this structure:

`{(resource_id, node_name): tags}`

This let's me use the same dictionary structure for AWS and Azure, but to unpack all the information and still get the right tag matching I just put the instance_id and tag generation together.

Reply:
Ok. Can we document that structure in the function docstring then? I think that's a good thing to capture for future readers. :smile: